File: 338.counting-bits.py
#
# @lc app=leetcode id=338 lang=python
#
# [338] Counting Bits
#

# @lc code=start
class Solution(object):
    def countBits(self, n):
        """
        :type n: int
        :rtype: List[int]
        """

       #DP
        # res = [0] * (n+1)
        # for i in range(1, n+1):
        #     res[i] = res[i >>1] + (i&1)

        # return res

        # Brute Force
        def count_ones(x):
            """
            Counts the number of '1' bits in the binary representation of an integer x.
            This is a helper function for the brute force approach.
            """
            count = 0
            while x:
                # Check the Least Significant Bit using bitwise AND with 1.
                # If the LSB is 1, 'x &1' will be '1', and we add it to the count
                # count = count + (x&1)
                count += x &1
                # After processing the LSB, right shift x by 1 to remove the LSB.
                # x = x >> 1
                x >>= 1
            return count
        
        return [count_ones(i) for i in range(n+1)]
        
        
# @lc code=end

File: 338.counting-bits.cpp
/*
 * @lc app=leetcode id=338 lang=cpp
 *
 * [338] Counting Bits
 */

// @lc code=start
#include <vector>

using namespace std;

class Solution
{

public:
        vector<int> countBits(int n)
    {
        // DP
        // All elements are initially set to 0. ans[0] is set to 0. This allows the loop to start iterating from i = 1
        // vector<int> ans(n + 1, 0);
        // for (int i = 1; i <= n; i++)
        // {
        //     // i = 5 → 二進位 = 101
        //     // i >> 1 = 2 → 二進位 = 10
        //     // countBits[5] = countBits[2] + 1
        //     //  ans[i] = ans[i / 2] + (i % 2)
        //     ans[i] = ans[i >> 1] + (i & 1);
        // }
        // return ans;

        // Brute Force
        vector<int> ans(n + 1, 0);
        for (int i = 1; i <= n; i++)
        {
            ans[i] = count_ones(i);
        }
        return ans;
    }
    int count_ones(int x)
    {
        int count = 0;
        while (x)
        {
            count += x & 1;
            x >>= 1;
        }
        return count;
    }
};
// @lc code=end
